packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/BoundedNonWhitespaceSequenceScopeHandler.ts

```diff
@@ -0,0 +1,178 @@
+  ): Iterable<TargetScope> {
+    const targetScopes = this.targetScopeHandler.generateScopes(
+      editor,
+      position,
+      direction,
+      {
+        ...hints,
+        // Don't skip containing paint since it might have non contained nested scopes.
+        containment:
+          hints.containment !== "disallowed" ? hints.containment : undefined,
+      },
+    );
+
+    const pairScopes = Array.from(
+      this.surroundingPairInteriorScopeHandler.generateScopes(
+        editor,
+        position,
+        direction,
+        {
+          ...hints,
+          // For every (skipAncestorScopes=true) we don't want to go outside of the surrounding pair
+          containment:
+            hints.containment == null && hints.skipAncestorScopes
+              ? "required"
+              : hints.containment,
+        },
+      ),
+    );
+
+    for (const targetScope of targetScopes) {
+      const allScopes = [targetScope];
+
+      for (const pairScope of pairScopes) {
+        const interiorRange = pairScope.domain;
+        const intersection = targetScope.domain.intersection(interiorRange);
+        if (intersection != null && !intersection.isEmpty) {
+          allScopes.push(this.createTargetScope(editor, intersection));
+        }
+      }
+
+      allScopes.sort((a, b) => compareTargetScopes(direction, position, a, b));
+
+      yield* allScopes;
+    }
+  }
+
+  protected abstract createTargetScope(
+    editor: TextEditor,
+    contentRange: Range,
+  ): TargetScope;
+}
+
+export class BoundedNonWhitespaceSequenceScopeHandler extends BoundedBaseScopeHandler {
+  public readonly scopeType = { type: "boundedNonWhitespaceSequence" } as const;
+
+  constructor(
+    scopeHandlerFactory: ScopeHandlerFactory,
+    scopeType: ScopeType,
+    languageId: string,
+  ) {
+    super(scopeHandlerFactory, languageId, { type: "nonWhitespaceSequence" });
+  }
+
+  protected createTargetScope(
+    editor: TextEditor,
+    contentRange: Range,
+  ): TargetScope {
+    return {
+      editor,
+      domain: contentRange,
+      getTargets(isReversed) {
+        return [
+          new TokenTarget({
+            editor,
+            isReversed,
+            contentRange,
+          }),
+        ];
+      },
+    };
+  }
+}
+
+export class BoundedParagraphScopeHandler extends BoundedBaseScopeHandler {
+  public readonly scopeType = { type: "boundedParagraph" } as const;
+
+  constructor(
+    scopeHandlerFactory: ScopeHandlerFactory,
+    scopeType: ScopeType,
+    languageId: string,
+  ) {
+    super(scopeHandlerFactory, languageId, { type: "paragraph" });
+  }
+
+  protected createTargetScope(
+    editor: TextEditor,
+    contentRange: Range,
+  ): TargetScope {
+    return {
+      editor,
+      domain: contentRange,
+      getTargets(isReversed) {
+        return [
+          new ParagraphTarget({
+            editor,
+            isReversed,
+            contentRange,
+          }),
+        ];
+      },
+    };
+  }
```

Comment:
I wonder how things work in this case. Because paragraph targets kind of expect to actually be a full block. See the test cases I added